### ISSUE TITLE
DNS 'server misbehaving' with SPF and DMARC checks (suspected Go bug)

### ISSUE THREAD

**Issue by reivilibre** [foxcpp/maddy#311]
When I send an e-mail to my server from, uhhh, 'example.net', Maddy Mail Server temporarily rejects it — keywords are that the lookup returns 'server misbehaving'.

In the logs, I see

```
smtp: RCPT ok        {"msg_id":"cc89eb1d","rcpt":"hello@mydom.net"}
msgpipeline: no check action        {"check":"verify_dkim","msg_id":"cc89eb1d","reason":"No DKIM signatures","smtp_code":550,"smtp_enchcode":"5.7.20","smtp_msg":"No DKIM signatures"}
check.spf: DMARC fetch        {"from_domain":"example.net","msg_id":"cc89eb1d","reason":"lookup _dmarc.example.net on 127.0.0.53:53: server misbehaving"}
smtp: DATA error        {"check":"check.spf","msg_id":"cc89eb1d","reason":"lookup example.net on 127.0.0.53:53: server misbehaving","smtp_code":451,"smtp_enchcode":"4.7.23","smtp_msg":"SPF authentication failed with a temporary error"}
```

This seems similar to an old issue caused by an issue in Go itself: https://github.com/golang/go/issues/12712
However, that was supposedly fixed in Go 1.6...?

Note: `dig example.net txt` suggests that there are no TXT records on that domain.

### Environment
`maddy 0.5.0-dev15+ge4ad3bd linux/amd64 go1.14.3`
Ubuntu 20.04.1 (LTS)

### Workaround
Removing search domains (lines beginning with `search`) in `/etc/resolv.conf` and restarting Maddy seems to have worked around the problem.

-----

Sorry for *yet another* issue, but even if there's nothing sensible to do about this, I thought it might be nice to have this here so someone else can find it and save themselves half an hour.

**Comment by foxcpp** [comment]
Another note: It is possible to make maddy ignore SPF in case of "temporary errors":
```
check {
  ...
  spf {
    temperr_action ignore # default is reject
  }
  ...
}
```

**Comment by foxcpp** [comment]
Is there any difference in DNS response to `TXT example.net.` lookup and lookups with search domains appended? 

E.g. if you have `search example.org. example.com.` in resolv.conf, do all of `TXT example.net.`, `TXT example.net.example.org.`, `TXT example.net.example.com.` return the same result (e.g. NOERROR without records)?

**Comment by foxcpp** [comment]
I tried a few cases and failed to reproduce the issue. I will need more information as mentioned in the previous post.

**Comment by foxcpp** [comment]
I just discovered that TestCheckSPF integration test starts to fail if I add `search example.org example.com` to resolv.conf.

**Comment by foxcpp** [comment]
This might be a problem in maddy code. Unfortunately, while RFCs require domains in records and in SMTP exchanges to be FQDNs they are almost always not formatted as ones (with trailing dot). Adding `search` to resolv.conf makes stub resolver interpret non-FQDN names in a different way (by prepending search domains instead of `.`) and results in wrong queries being sent. I believe forcing FQDN interpretation by adding `.` to all domains would be appropriate.

CC @albertito  I think chasquid may also have a similar problem. WDYT?

**Comment by foxcpp** [comment]
> I just discovered that TestCheckSPF integration test starts to fail if I add `search example.org example.com` to resolv.conf.

However, the error message is "no such host" in this case so that's likely a different issue.

**Comment by foxcpp** [comment]
@reivilibre the referenced commit should fix (or rather, mask) the issue you are having by forcing verified domains to be interpreted as FQDN, therefore ignoring `search` in resolv.conf.

I still would like to get more information as mentioned in https://github.com/foxcpp/maddy/issues/311#issuecomment-739434264 to troubleshoot the underlying cause. Also, what DNS server are you using?

**Comment by reivilibre** [comment]
Sorry for the delay,

I think I've found the issue.

A search domain of `example.org` works fine because `example.net.example.org.` gives a NXDOMAIN and `example.net.` gives the correct response.

However, my resolv.conf had the search domain `invalid` in it for some reason (I don't know why... It regenerated from netplan since, but I've taken that out now). Digging `example.net.invalid.` gives `SERVFAIL` which is probably the reason this gives 'server misbehaving' errors.

I will try your new commit shortly and verify that it works, however I have no reason to believe it won't.

The DNS server I'm using is systemd-resolved on `127.0.0.53` (not that I like it... had plenty of issues) which then resolves from some others.

**Comment by hibobmaster** [comment]
@reivilibre  Thank you very much. You solved my problem that has annoyed me for three days.
I have test postfix+dovecot、docker-mailserver、chasquid and finally find the solution here,
😀

**Comment by reivilibre** [comment]
> @reivilibre Thank you very much. You solved my problem that has annoyed me for three days.
> I have test postfix+dovecot、docker-mailserver、chasquid and finally find the solution here,
> grinning

Glad to hear it, good to know my plan of posting about this to hopefully help others has already been successful :)

**Comment by foxcpp** [comment]
Closing as this is fixed in d6a0324.
